feat: Track the Latest Schedule Version on a User's Device (M2-8478)

sultanofcardio opened the pull request:
- [ ] Tests for the changes have been added

### 📝 Description

🔗 [Jira Ticket M2-8478](https://mindlogger.atlassian.net/browse/M2-8478)

> [!NOTE]
> This PR is based on https://github.com/ChildMindInstitute/mindlogger-backend-refactor/pull/1723. If that is still open, please review it instead

This PR creates the `user_device_events` table, and updates the API code to keep track of the latest schedule version that a particular user device downloads. It does this by intercepting requests to `/users/me/respondent/current_events`, which the mobile app uses to download the latest schedules for a user, and inserting/updating records in this new table

| Column Name        | Data Type          | Constraints                                                      | Description                                       |
|--------------------|--------------------|------------------------------------------------------------------|---------------------------------------------------|
| `is_deleted`       | Boolean            | Nullable                                                         | Indicates if the record is soft deleted. Not used |
| `created_at`       | DateTime           | Nullable, Default: `timezone('utc', now())`                      | Timestamp when the record was created             |
| `updated_at`       | DateTime           | Nullable, Default: `timezone('utc', now())`                      | Timestamp when the record was last updated        |
| `migrated_date`    | DateTime           | Nullable                                                         | Not used                                          |
| `migrated_updated` | DateTime           | Nullable                                                         | Not used                                          |
| `id`               | UUID               | Primary Key, Nullable                                            | Unique identifier for the record                  |
| `device_id`        | UUID               | Foreign Key: `user_devices.id`, Not Nullable, On Delete: CASCADE | UUID foreign key to the user_devices table        |
| `event_id`         | UUID               | Foreign Key: `events.id`, Not Nullable, On Delete: CASCADE       | UUID foreign key to the events table              |
| `version`          | String (length=13) | Not Nullable                                                     | Most recent downloaded version of a given event   |

**Unique Constraint**: `device_id`, `event_id`

### 🪤 Peer Testing

- Run the migration
    ```shell
    alembic upgrade head
    ```
- Confirm that the table has been created and is empty
- Start up the API and launch the mobile app connected to your local API. Please use [this mobile app PR](https://github.com/ChildMindInstitute/mindlogger-app-refactor/pull/932) to test it.
- Swipe refresh the app
- Check the table again and confirm there are the same number of entries are there are activities and flows in the applets your user has access to. You can do this by checking the table directly, the running the query below and confirming the number of records are the same

    <img width="1703" alt="image" src="https://github.com/user-attachments/assets/1267d8e9-1b5c-4c61-acc2-1d75f8e1fc77" />

    ```sql
    SELECT ude.device_id,
           ude.event_id,
           ude.version as event_version,
           a.display_name as applet,
           COALESCE(ac.name, f.name) as activity_or_flow
    FROM user_device_events ude
    JOIN events e ON ude.event_id = e.id
    JOIN applets a ON e.applet_id = a.id
    LEFT JOIN activities ac ON a.id = ac.applet_id
    LEFT JOIN flows f ON a.id = f.applet_id;

    ```
    <img width="1520" alt="image" src="https://github.com/user-attachments/assets/fe6c7d77-c028-41c9-8a44-1fc8cb436fdd" />
- Delete the records in the table and repeat the test with the version of the mobile app from the `dev` branch. The table should remain empty and there should be no API or mobile app errors

### ✏️ Notes

N/A

rcmerlo commented on src/apps/schedule/crud/user_device_events.py:
```diff
@@ -0,0 +1,43 @@
+        ]
+
+        upsert: Insert = pg_insert(UserDeviceEventsSchema)
+        upsert = upsert.values(values)
+        upsert = upsert.on_conflict_do_update(
+            constraint=UserDeviceEventsSchema.unique_constraint,
+            set_={
+                "version": upsert.excluded.version,
+                "updated_at": datetime.datetime.now(datetime.UTC).replace(tzinfo=None),
+            },
+        )
+        upsert = upsert.returning(UserDeviceEventsSchema)
```
```suggestion
        upsert.values(values)
        upsert.on_conflict_do_update(
            constraint=UserDeviceEventsSchema.unique_constraint,
            set_={
                "version": upsert.excluded.version,
                "updated_at": datetime.datetime.now(datetime.UTC).replace(tzinfo=None),
            },
        )
        upsert.returning(UserDeviceEventsSchema)
```

All functions here, returns None, so don't need to assign to upsert

rcmerlo commented on src/apps/schedule/crud/user_device_events.py:
```diff
@@ -0,0 +1,43 @@
+        upsert = upsert.returning(UserDeviceEventsSchema)
+        result = await self._execute(upsert)
+
+        rows = result.mappings().all()
+        model = [UserDeviceEventsSchema(**row) for row in rows]
+
+        return model
```
```suggestion
        rows = result.scalars().all()

        return rows
```

We are returning UserDeviceEventsSchema, so better to use scalars().all(), it returns a list of UserDeviceEventsSchema objects

sultanofcardio replied:
I'm not sure why this doesn't work, but when I try this I just get a list of `bool`s

<img width="616" alt="image" src="https://github.com/user-attachments/assets/30b3c6ba-b716-4671-8274-811524f04aaa" />

<img width="618" alt="image" src="https://github.com/user-attachments/assets/1eff0eeb-0862-4b92-9eb4-3607f9210748" />

I assume these correspond to which rows were updated because they're `False` when the rows are inserted

rcmerlo reviewed (commented):
Few suggestions

sultanofcardio commented:
@rcmerlo this PR needs a re-review. I've done the following things since you last reviewed:
- I renamed the table `user_device_events_history`
- I added the `version` column to the unique constraint to allow for keeping track of each version of each event the user downloads, not just the latest one
- I removed the foreign key constraints from `device_id` and `event_id`. This is because `user_devices` records don't live forever, but these history records need to live forever. There may also be records in this table that references events that no longer exist
- I added a `user_id` column for ease of reading that **is** a foreign key to the `users` table

rcmerlo commented:
Can you add some tests?

sultanofcardio commented:
@rcmerlo added some tests
